### ISSUE TITLE
Client Streaming hangs if server returns a gRPC error

### ISSUE THREAD

**Issue by mwitkow** [grpc/grpc-go#764]
We want to use gRPC Client (many) -> Server (single) streaming:

```
service LogSinkService{
    rpc LogStream(stream LogEntry) returns (google.protobuf.Empty);
}
```

We want to do auth on establishing this Stream and after just log the stream of messages. Basically compare a field of the first message in the stream against a `Bearer` token in the Metadata of the call using our permission system. 

We use gRPC error codes to signal problems with our calls in a uniform fashion. We have Logging and Monitoring Interceptors that take the error codes of responses and act on them. This allows us to have a uniform inspection and alerting mechanisms regardless of payloads.

For auth we use `PermissionDenied` (code 7), and would like us to be able to respond to the user stream with that error code if the logging request is unauthorized.

However, we hit two issues:

## Issue 1
There is no way for a sender to know the status from the server prior to doing `ReceiveAndClose`. 

Basically if a client is doing:

```go
stream, err := client.LogStream(ctx)
sendErr := stream.Send(newLogMessage())
endErr := stream.ReceiveAndClose()
```
the gRPC server's error code (in our case `PermissionDenied`) is not returned in `sendErr`. It just happily write it down the connection. It would be helpful to have the error potentially returned on first attempt to send, but I understand that the underlying protocol spec doesn't allow for that.

That is fine, we'd be fine shoving some bytes down the wire, but then we hit another issue.

## Issue 2
The client runs out of `sendQuotaPool` but doesn't receive a response.

Basically if you do:

```go
stream, err := client.LogStream(ctx)
for i := 0; i < 10000; i++ {
  sendErr := stream.Send(newLogMessage())
}
endErr := stream.ReceiveAndClose()
```

at some point the client just hangs forever until a Context Deadline Exceeded (if you set the timeline). Basically it seems that `http2Client` waits on `sendQuotaPool` without knowing that the server already terminated the connection with an error. 

src/google.golang.org/grpc/transport/http2_client.go:471
```go
func (t *http2Client) Write(s *Stream, data []byte, opts *Options) error {
			// Wait until the stream has some quota to send the data.
....
			sq, err := wait(s.ctx, t.shutdownChan, s.sendQuotaPool.acquire())
````
The `shutdownChan` doesn't get invoked because nobody closed the channel (nobody read the response from the server) and there is noone to replenish the `sendQuotaPool` because the other server didn't call `Recv` (as it just returned from the call).

I think that this doesn't affect the bidirectional streaming because most users send a request and expect a response, thus the `sendQuotaPool` gets replenished. However for Client to Server streaming it may be a problem.

**Comment by iamqizhao** [comment]
Yes, some users already reported this. The fix should be out soon.

**Comment by mwitkow** [comment]
@iamqizhao great to hear. 

Will the fix involve returning grpc Code errors as part of `.Send()` so that the client can see that the server doesn't accept any more traffic.
